README.md
Librerie per la realizzazione di gestionali web-based in PHP. In italiano.

changelog.txt
2016-03-11 tag rel_3_1_5
- modificato banner

2014-11-28 tag rel_3_1_4
- corretti driver mysql: c'era un warning al ritorno di una scrittura andata 
bene (cercava di fetchare il risultato)

2014-11-13 tag rel_3_1_3
- corretto caricafile.xsl (erano rimasti dei doppi apici dannosi)

2014-06-25 tag rel_3_1_2
- corretta gestione decimali valuta

2014-05-16 tag rel_3_1_0
- cdnizzazione: rimossa (fin dove possibile) la presenza di librerie terze parti,
	sia php che js:
	- php: 
		- phpmailer e fpdf devono trovarsi dentro direcory così chiamate in include_path
		- analogo al punto precedente per PHPExcel e ExcelWriterXML
		- analogo ai punti precedenti per i moduli PEAR (Spreadsheet)
	- js: tutto cio' che è terze parti è in CDN, ad eccezione della gestione
			del calendario delle date e di ddmx: è roba troppo vecchia che andrà
			presto sostituita

2014-02-23 tag rel_3_0_28
- corretto possibilità di special chars in menu e titolo tabella

2014-02-23 tag rel_3_0_27
- aggiunto gestione tipo "newdate"....

2014-02-20 tag rel_3_0_26
- hanno corretto un baco in mysql (utf -> len x 3) per cui ho tolto il workaround relativo in wadocumentazione

2014-02-19 tag rel_3_0_25
- aggiunta possibilita, in fase di edit, aggiungere la riga in fondo oltreche all'inizio della tabella

2014-02-10 tag rel_3_0_24
corretti 2 errori (2!) che impedivano alla tabella di avere la riga dei totali partendo da vuota

2014-02-10 tag rel_3_0_23
- piccola correzione su bottone

2014-01-19 tag rel_3_0_22
- aggiunti (o modificati) metodi magic su warecord e wamodulo
- corretto errori su multiselezione_checkbox
- waselezione: ordinamento lista di default su riempimento rpc (per valore e non per chiave)
- wavaluta: corretto punto che non diventava virgola (in tempo)
- watabella: portato a fattor comune un po' di template xsl
- watabella: corretto errore per cui resettava parametri qs su ricerca rapida
- watabella: cambiato comportamento di pagina precedente/successiva (ma mantiene compatibilità)

2014-01-10 tag rel_3_0_21
- corretto errore su formattazione wavaluta
- aggiunte proprietà automagic a waRecord

2014-01-01 tag rel_3_0_20
- corretto piccoli errori e cambiato metodo larghezza di default delle colonne 
	in esportazione pdf di watabella

2013-11-22 tag rel_3_0_19
- aggiunta gestione waApplicazione::aggiungiElemento XML

2013-11-08 tag rel_3_0_18
- aggiunti tag xml versione librerie xslt 
- corretto errori su driver MySQLi

2013-10-30 tag rel_3_0_17
- aggiunti tag xml relativi alla sezione in waApplicazione

2013-10-25 tag rel_3_0_16
- aggiunta gestione eventi su UI di multiselezione_checkbox

2013-09-17 tag rel_3_0_14
- modificata mandaMail : gestisce exception applicative (altrimenti la versione 5.2 di phpMailer faceva un echo a fronte di exception applicativa... strana gestione...)

2013-08-02 tag rel_3_0_14
- corretto su waDb: velocizzata cancellazione da recordset

2013-08-02 tag rel_3_0_13
- corretto su waDb: errore presente dal 2008, tosy's rulez: la cancellazione 
di più di un record contemporaneamente provocava la cancellazione di un solo
record (il primo che soddisfaceva la condizione)

2013-07-30 tag rel_3_0_12
- corretto su waTabella : strano errore su costruzione elemento URI: 
	http_build_query faceva un encoding non gradito. Ma l'ha sempre fatto?
	Adesso comunque sembra a posto
- corretto su waTabella: errore su tipo input logico e quick-edit
- waDB: operatore eq dannoso su recid in FM 12
- waDocumentazione può aggiungere elementi


2013-07-05 tag rel_3_0_11
- corretto su waModulo : elimina() elimina solo se il record ha un solo
	campo chiave primaria

2013-07-01 tag rel_3_0_10
- reverting completo sulle date: si usano solo date e mktime (no gm...);
se si incorre nuovamente nel problema del dst sbagliato per date vecchie (
1904-01-01
1944-08-21
1945-07-16
1941-11-09
)
allora faremo un nuovo controllo che salva il campo stringa iso

2013-07-01 tag rel_3_0_9
- nessuna modifica: scordato solamente il changelog...

2013-07-01 tag rel_3_0_8
- wamodulo: corretto date -> gmdate in modulo.class.php
- watabella: inclusione in CDATA di un valore dì una colonna per cui è stato richiesto convertiHTML = false

2013-06-28 tag rel_3_0_7
- errore su creazione oggetti javascript


2013-06-20 tag rel_3_0_6
- metodo mostra di waApplicazione; correggiHTML per conformità strict a XHTML 
1.0 chiamata solo in caso di effettivo html e di pagina intera

2013-06-12 tag rel_3_0_5
- aggiunto controllo captcha a waModulo + aggiornata documentazione

2013-06-11 tag rel_3_0_4
- aggiunto controllo cfpi a waModulo

2013-06-10 tag rel_3_0_3
- errore su metodo mostra di waApplicazione; veniva chiamata 2 volte correggiHTML
	per conformità strict a XHTML 1.0

2013-03-27
- tutte le funzioni data fanno riferimento sempre a GMT
- eliminato adodb

2013-03-25
- aggiunto waNonControllo
- aggiunto multiselezione_checkbox (xslt)


wamodulo/cfpi.class.php
<?php

/**
 * @package waModulo
 * @version 3.0
 * @author G.Gaiba, F.Monti
 * @copyright (c) 2007-2016 {@link http://www.webappls.com WebAppls} Bologna, Italy
 * @license http://www.gnu.org/licenses/gpl.html GPLv3
 */
if (!defined('_WA_CFPI'))
	{
	/**
	 * @ignore
	 */
	define('_WA_CFPI', 1);

	/**
	 * @ignore
	 */
	include_once(dirname(__FILE__) . "/testo.class.php");

//***************************************************************************
//****  classe waCFPI *******************************************************
//***************************************************************************
	/**
	 * waCFPI
	 *
	 * classe per la gestione dei controlli destinati a contenere un codice fiscale,
	 * una partita IVA o entrambi. 
	 * 
	 * E' un normale {@link waTesto} dal quale si differenzia solo per il tipo, in 
	 * modo da permetterne il riconoscimento lato client e di conseguenza le
	 * relative procedure di controllo
	 * 
	 * @package waModulo
	 * @version 3.0
	 * @author G.Gaiba, F.Monti
	 * @copyright (c) 2007-2016 {@link http://www.webappls.com WebAppls} Bologna, Italy
	 * @license http://www.gnu.org/licenses/gpl.html GPLv3
	 */
	class waCFPI extends waTesto
		{

		/**
		 * @ignore
		 * @access protected
		 */
		var $tipo = 'cfpi';

		/**
		 * flag di gestione CF (default true)
		 * @var boolean
		 */
		var $gestioneCF = true;

		/**
		 * flag di gestione PI (default false)
		 * @var boolean
		 */
		var $gestionePI = false;

		//***************************************************************************
		/**
		 * @access protected
		 * @ignore
		 */
		function mostra()
			{
			if ($this->gestioneCF)
				$this->caratteriMax = $this->caratteriVideo = 16;
			elseif ($this->gestionePI)
				$this->caratteriMax = $this->caratteriVideo = 11;
			
			$this->xmlOpen();
			$this->xmlAdd("valore", $this->valore);
			$this->xmlAdd("caratteri_max", $this->caratteriMax);
			$this->xmlAdd("caratteri_video", $this->caratteriVideo);
			$this->xmlAdd("gestione_cf", $this->gestioneCF);
			$this->xmlAdd("gestione_pi", $this->gestionePI);
			$this->xmlClose();
			}

		//****************************************************************************************
		/**
		 * Restituisce il valore se valido, altrimenti null
		 *
		 * Si usa in fase di ricezione dei dati, non
		 * durante la costruzione della form.
		 *
		 * @ignore
		 * @return string
		 */
		function input2valoreInput($valoreIn)
			{
			if ($valoreIn === "__wamodulo_valore_non_ritornato__")
				{
				$this->inputNonRitornato = true;
				return null;
				}

			if ($valoreIn === '') return null;

			$tester = strtoupper(substr($valoreIn, 0, 1));
			if (preg_match('/^[A-Z]$/', $tester) && $this->gestioneCF) return $this->verificaCF($valoreIn);
			elseif (preg_match('/^\d$/', $tester) && $this->gestionePI) return $this->verificaPI($valoreIn);

			return null;
			}

		//****************************************************************************************
		/**
		 * @ignore
		 * @return string
		 */
		protected function verificaCF($valoreIn)
			{
			$pari = array(
				'0' => 0, '1' => 1, '2' => 2, '3' => 3, '4' => 4,
				'5' => 5, '6' => 6, '7' => 7, '8' => 8, '9' => 9,
				'A' => 0, 'B' => 1, 'C' => 2, 'D' => 3, 'E' => 4,
				'F' => 5, 'G' => 6, 'H' => 7, 'I' => 8, 'J' => 9,
				'K' => 10, 'L' => 11, 'M' => 12, 'N' => 13, 'O' => 14,
				'P' => 15, 'Q' => 16, 'R' => 17, 'S' => 18, 'T' => 19,
				'U' => 20, 'V' => 21, 'W' => 22, 'X' => 23, 'Y' => 24,
				'Z' => 25
			);

			$dispari = array(
				'0' => 1, '1' => 0, '2' => 5, '3' => 7, '4' => 9,
				'5' => 13, '6' => 15, '7' => 17, '8' => 19, '9' => 21,
				'A' => 1, 'B' => 0, 'C' => 5, 'D' => 7, 'E' => 9,
				'F' => 13, 'G' => 15, 'H' => 17, 'I' => 19, 'J' => 21,
				'K' => 2, 'L' => 4, 'M' => 18, 'N' => 20, 'O' => 11,
				'P' => 3, 'Q' => 6, 'R' => 8, 'S' => 12, 'T' => 14,
				'U' => 16, 'V' => 10, 'W' => 22, 'X' => 25, 'Y' => 24,
				'Z' => 23
			);

			$valoreIn = strtoupper($valoreIn);
			$sum = 0;

			for ($i = 1; $i < 16; $i++) $sum += ($i % 2) ? $dispari[$valoreIn[$i - 1]] : $pari[$valoreIn[$i - 1]];

			if (chr($sum % 26 + ord('A')) != substr($valoreIn, 15))
			// check digit sbagliato return null;

			if (preg_match('/^[A-Z]{6}\d{2}[A-EHLMPRST]\d{2}[A-Z]\d{3}[A-Z]$/', $valoreIn))
			// codice standard, non c'e' bisogno di controllare omocodia return $valoreIn;

			// verifichiamo che la parte letterale sia composta effettivamente di
			// lettere e che la parte normalmente numerica contenga solo lettere
			// previste nei casi di omocodia
			if (!preg_match('/^[A-Z]{6}[\dL-NP-V]{2}[A-EHLMPRST][\dL-NP-V]{2}[A-Z][\dL-NP-V]{3}[A-Z]$/', $valoreIn)) return null;

			// una eventuale lettera al posto del numero indicante la decina del mese
			// non può essere una U o una V (corrispondenti a 8 e 9)
			if (!preg_match('/^[\dA-Z]{9}[\dL-NP-T][\dA-Z]{6}$/', $valoreIn)) return null;

			return $valoreIn;
			}

		//****************************************************************************************
		/**
		 * controlla una pèartita iva
		 * @ignore
		 * @return string
		 */
		protected function verificaPI($valoreIn)
			{
			//la p.iva deve avere solo 11 cifre
			if (!preg_match('/^[0-9]{11}$/', $valoreIn)) 
				return null;

			$primo = 0;
			for ($i = 0; $i <= 9; $i+=2) $primo+= ord($valoreIn[$i]) - ord('0');

			for ($i = 1; $i <= 9; $i+=2)
				{
				$secondo = 2 * ( ord($valoreIn[$i]) - ord('0') );
				if ($secondo > 9) 
					$secondo = $secondo - 9;
				$primo+=$secondo;
				}
			if ((10 - $primo % 10) % 10 != ord($valoreIn[10]) - ord('0'))
				return null;

			return $valoreIn;
			}

		//****************************************************************************************
		}

	// fine classe waCFPI
//***************************************************************************
//******* fine della gnola **************************************************
//***************************************************************************
	} //  if (!defined('_WA_CFPI'))
?>